src/main/java/com/auth0/TransientCookieStore.java

```diff
@@ -0,0 +1,154 @@
+
+    // Prevent instantiation
+    private TransientCookieStore() {}
+
+
+    /**
+     * Stores a state value as a cookie on the response.
+     *
+     * @param response the response object to set the cookie on
+     * @param state the value for the state cookie. If null, no cookie will be set.
+     * @param sameSite the value for the SameSite attribute on the cookie
+     * @param useLegacySameSiteCookie whether to set a fallback cookie or not
+     */
+    static void storeState(HttpServletResponse response, String state, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        store(response, StorageUtils.STATE_KEY, state, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Stores a nonce value as a cookie on the response.
+     *
+     * @param response the response object to set the cookie on
+     * @param nonce the value for the nonce cookie. If null, no cookie will be set.
+     * @param sameSite the value for the SameSite attribute on the cookie
+     * @param useLegacySameSiteCookie whether to set a fallback cookie or not
+     */
+    static void storeNonce(HttpServletResponse response, String nonce, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        store(response, StorageUtils.NONCE_KEY, nonce, sameSite, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the state cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the state cookie, if it exists
+     */
+    static String getState(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.STATE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    /**
+     * Gets the value associated with the nonce cookie and removes it.
+     * @param request the request object
+     * @param response the response object
+     * @param useLegacySameSiteCookie whether to use a fallback cookie or not
+     * @return the value of the nonce cookie, if it exists
+     */
+    static String getNonce(HttpServletRequest request, HttpServletResponse response, boolean useLegacySameSiteCookie) {
+        return getOnce(StorageUtils.NONCE_KEY, request, response, useLegacySameSiteCookie);
+    }
+
+    private static void store(HttpServletResponse response, String key, String value, SameSite sameSite, boolean useLegacySameSiteCookie) {
+        Validate.notNull(response, "response must not be null");
+        Validate.notNull(key, "key must not be null");
+        Validate.notNull(sameSite, "sameSite must not be null");
+
+        if (value == null) {
+            return;
+        }
+
+        boolean sameSiteNone = SameSite.NONE == sameSite;
+
+        String cookie = String.format("%s=%s; HttpOnly; Max-Age=%d; SameSite=%s", key, value, MAX_AGE_SECONDS, sameSite.getValue());
```

Comment:
Could the `key` or `value` here include a character that will break the header (semicolon)?

Reply:
Great catch. Short is answer is yes. The key is not a big concern since it is only set internally. But the state and nonce values can be set by the developer (by default we will generate a Base 64 encoded value, but it can be set to any string if the developer chooses). Will need to evaluate how to handle, perhaps we encode/decode the cookie values.

Reply:
It might be possible to escape the value ... I don't know OTOH if that's possible. Or just reject values that can break it (semi-colon and ... ?). I don't think there's anything wrong with a failure here; invalid chars set by the developer here feels very edge case.

Reply:
Yes, does feel edge case since by default we create the nonce and state values, but they can be customized if the developer chooses to. Will look at what chars we need to reject or if we should escape the values.

Reply:
Updated to encode the values, which handles problematic characters.